### PR TITLE
Remove Phobos 3 dub.sdl and replace it with the build_v3.d build script.

### PR DESCRIPTION
**Pull request by LightBender**
The DUB idea wasn't cutting it. So we decided to replace it with a build script written in D. Henceforth `dmd -run build_v3.d unittest` will now build and test Phobos 3.

### CODE REVIEW

**Comment by dkorpel** [comment]
> The DUB idea wasn't cutting it. So we decided to replace it with a build script written in D.

I wasn't part of that discussion, but that sounds like either dub has a big problem, or the build process is overly complex. Neither of these is desirable, so I'd rather change those than improvise yet another build system. dmd's build.d script has been a headache, and if we have to go that route again, at least document exactly what the build script does, and what it does differently than dub/make.

**Comment by LightBender** [comment]
> > The DUB idea wasn't cutting it. So we decided to replace it with a build script written in D.
> 
> I wasn't part of that discussion, but that sounds like either dub has a big problem, or the build process is overly complex. Neither of these is desirable, so I'd rather change those than improvise yet another build system. dmd's build.d script has been a headache, and if we have to go that route again, at least document exactly what the build script does, and what it does differently than dub/make.

It was @jmdavis who was having issues, so I'll let him fill you in on the specifics.

**Comment by dkorpel** [comment]
With GitHub actions, you can get dub on the build machine with the dlang-community/setup-dlang action. Example from dmd:
https://github.com/dlang/dmd/blob/master/.github/workflows/build_with_dub.yml

**Comment by jmdavis** [comment]
1. We're having issues with the CI, because dub isn't working with it. The result is that right now, Phobos v3 is not tested at all by the CI.
2. It would be nice to make it so that we can build and run subsets of the unit tests like the makefile does. That's not needed yet, since Phobos v3 isn't large yet, but it will matter later, and that's not something that dub can sanely do.
3.  However, the big reason is simply that what we ultimately need to do here is replace Phobos' makefile, and it does stuff like build the documentation, which isn't going to be even vaguely sane to do with dub. I don't see how it makes any sense to use dub for the Phobos build in the long run. It's just too simplistic, and it's a royal pain to deal with (if not impossible) when you want to do anything that isn't incredibly simple. And there is no need to put the standard library up on code.dlang.org or to pull anything from code.dlang.org in order to build the standard library, which are the main reasons to use dub with any project. Getting rid of dub also removes a build dependency, and we want to minimize those with the standard library. I see _no_ advantage whatsoever in using dub for Phobos.

The only semi-sane options I see here are to either use a makefile or to use a script. A D script is going to be the easiest way to do a script, and it's going to be much easier to deal with than a makefile.

Right now, this is a build script specific to Phobos 3, but I would expect that by the time we're looking to release Phobos v3, we would have reworked the Phobos build so that it's all done with a build script written in D and not the makefile.

I can't speak to the specifics of what Adam has done, since I haven't reviewed it yet, but the script can be adjusted as needed. The key thing at the moment is to just get a basic build working with the ability to run tests without involving dub, since it's just getting in the way and clearly won't be able to handle what we'll need to do in the long run.

**Comment by dkorpel** [comment]
I gather you view Phobos like a behemoth that can't possibly be tamed by the measly 'dub', unlike other libraries on code.dlang.org. While Phobos is extensive, it's qualitatively no different than other D libraries in my view. Phobos should set an example. The example we want to set is "D libraries are simple and our official build tool dub is production ready". 
Not "D libraries are a complex tangle, and our tool is inadequate for that so you better maintain your own build script in a turing complete language instead".

> 1. We're having issues with the CI

There's `dlang-community/setup-dlang` which includes dub

> 2. It would be nice to make it so that we can build and run subsets of the unit tests like the makefile does. that's not something that dub can sanely do.

Yes it can, you can provide arguments to a custom test runner like `dub test -- --include=std.algorithm`. See [silly](https://code.dlang.org/packages/silly) for a real example.

> 3. However, the big reason is simply that what we ultimately need to do here is replace Phobos' makefile, and it does stuff like build the documentation, which isn't going to be even vaguely sane to do with dub. 

When dub can't build documentation for Phobos, it can't build it for any library, which needs to be fixed.

> It's just too simplistic, and it's a royal pain to deal with (if not impossible) when you want to do anything that isn't incredibly simple. 

People have avoided Phobos v2 for its complexity. If Phobos 3 isn't incredibly simple, that's a failure.

> And there is no need to put the standard library up on code.dlang.org or to pull anything from code.dlang.org in order to build the standard library, which are the main reasons to use dub with any project.

The reason is standardization and simplicity. I want to tell new contributors: to start hacking on Phobos, just clone the repo and run the commands you know: `dub test`, `dub build` etc. I don't want to tell them: look at the documentation for which incantation does the simple obvious things. If there's advanced steps that can't be easily be done via dub, by all means add scripts for those tasks. But building and running tests should be simple.

> The key thing at the moment is to just get a basic build working with the ability to run tests without involving dub, since it's just getting in the way and clearly won't be able to handle what we'll need to do in the long run.

How about: we start using dub, and once we actually add something that hits a limitation of dub, we evaluate if it can be avoided, or if there's a deficiency in dub that needs to be fixed. If not, *then* to switch to a custom script.

**Comment by jmdavis** [comment]
Well, I just flat out disagree here. dub has a terrible build system, and in any situation where you need to do anything more complex than simply hand over a source directory and turn that into a library or an executable quickly becomes miserable. The Phobos build does more than that, and it also has to in order to deal with stuff like building druntime and pulling it in, because they share a library file. And I definitely wouldn't use dub to build documentation, particularly since how that build works is going to vary considerably depending on where you're putting it and in what form (it could be html, or it could be an epub file, or it could be a variety of things depending on how the build is set up).

Ideally, dub's current build system would be completely overhauled, and this issue does affect more than just Phobos, but given the current state of dub, I don't think that it makes any sense to try to use it unless you're explicitly looking to put your code up on code.dlang.org, or you need to pull in other libraries from code.dlang.org - neither of which Phobos needs.

What Phobos' Makefile currently does _might_ be possible to do with dub, but we'd have to fight it to get what we want and farm parts of the process out to scripts - or just bypass the whole thing with something like reggae. And if we want to do that, we might as well just avoid pulling in additional dependencies and just write a build script in D. Having additional build dependencies is less of an issue with other libraries, because they're not part of the bootstrapping process of getting a working D setup on your system, but with Phobos (and even moreso with dmd and druntime), we should be minimizing dependencies.

And sure, it might be nice to tell someone that they can just run `dub build` or `dub run`, but I don't think that it's at all a big deal to tell them to run something like `./build.d` or `./build.d unittest` instead - particularly when dub is a terrible build tool.

**Review by ljmf00 — COMMENTED** [review]
This is pretty much downgrading the build infrastructure... This is not so overly complex that dub can't handle.

**Comment by ljmf00** [comment]
> dub has a terrible build system

I completely agree. But your alternative option is not an upgrade, a script or makefiles, which I consider worse than dub.

File: build_v3.d
Comment (jmdavis) [inline build_v3.d:115]: This needs to run the `unittestExecutable` command from above or be changed to do `./unittest` on POSIX systems. Also, build.d should be made executable (which would presumably need to be done on a POSIX system, whereas I'm guessing that you're doing this on Windows). I can try to do a pull request to your repo to do that if you'd like.

File: build_v3.d
Comment (jmdavis) [inline build_v3.d:26]: This is fine for the moment, but we are going to want to change it so that the normal unittest build does both debug and release builds (which is what `make unittest` does) - as well as probably having a way to indicate that you want to do a unittest build that's only debug or only release. That's currently handled with `BUILD=debug` or `BUILD=release`.

But I think that we can make those improvements with a separate PR.

**Review by jmdavis — CHANGES_REQUESTED** [review]

**Comment by LightBender** [comment]
First, the `setup/dlang` action does not work for Phobos for some as yet unknown reason. We attempted to get it working with this PR: https://github.com/dlang/phobos/pull/8992 However, that PR broke the CI system for Phobs and was reverted here: https://github.com/dlang/phobos/pull/8999

Therefore, for the moment, we cannot use DUB in the Phobos build pipeline. Full stop. Which is a growing problem as @jmdavis continues to work on Ranges.

@dkorpel @ljmf00 Here is my proposal. Since both of you assert that DUB is both capable of handling Phobos 3 and is superior to build scripts and make files. May I suggest that you port the entire Phobos 2 build process over to DUB? The reason I suggest this is quite simple. The Phobos 3 build process will be extremely similar to Phobos 2 build process. Therefore, if you succeed, we can use your DUB build file for Phobos 2 and a template for Phobos 3 with minimal effort. If you do not, well, at least we learned something and remain unblocked in the mean time.

As such, we will merge this to unblock ourselves in the near team. And you can work on the Phobos 2 DUB build at your leisure. If you do successfully replace the existing Phobos 2 build scripts. I will personally merge your work.

**Comment by dkorpel** [comment]
Like ljmf00 said, it's not about dub being particularly good, it's about complex build scripts being terrible. I don't know if you had the pleasure of dealing with [dmd's build.d](https://github.com/dlang/dmd/blob/master/compiler/src/build.d), but I certainly have.

- It's 2300 lines of code, mixing the implementation of a build system and the configuration of the build. This makes it needlessly hard to modify, e.g. adding/removing a `-preview` flag.
- New commands you have to remember. It's not just `dub build` vs `./build.d`, it's also standard flags `--compiler=ldc --arch=x86 --build=release` now being environment variables `HOST_DMD=ldmd2 MODEL=32 ENABLE_RELEASE` etc.
- Non-standard output when a command fails, making it harder to read CI logs.
- There's bugs. E.g. it often doesn't rebuild the backend when it should, so I often have to do `./build.d clean` before building just to ensure nothing is corrupt.

No one wants to fix this, and even if someone did, that effort could have been directed towards improving dub: something which everyone benefits from instead of just one repository. An now the plan is to maintain *two* additional build systems. The only concrete reasons I've seen so far to justify this cost are:

> The Phobos build does more than that, and it also has to in order to deal with stuff like building druntime and pulling it in, because they share a library file.

> it could be html, or it could be an epub file

- Linking in druntime
- .epub documentation generation

So like I suggested: let's first evaluate if this is really necessary. Why does druntime have to be combined with Phobos? I always found it silly that I can't just clone the dmd repo, build it, and run its test: I also have to clone Phobos, build that, just so the tests have a druntime lib to link to.

Then if we decide to keep that complexity: let's identify what the exact command is that our custom build system would run to achieve it. Is that really something dub can't with the "sourceFiles" or "postBuildCommand" directives, or do we just assume dub is too limited?

If that doesn't work, perhaps we can have a custom script for that specific complex task, without re-inventing all the rest.

**Comment by LightBender** [comment]
> Like ljmf00 said, it's not about dub being particularly good, it's about complex build scripts being terrible. I don't know if you had the pleasure of dealing with [dmd's build.d](https://github.com/dlang/dmd/blob/master/compiler/src/build.d), but I certainly have.
> 
>     * It's 2300 lines of code, mixing the implementation of a build system and the configuration of the build. This makes it needlessly hard to modify, e.g. adding/removing a `-preview` flag.

I would suggest that a large amount of this code is spent dealing with the bits of the old build scripts that it still needs. Phobos 3 does not suffer from needing to interact with these legacy systems. The current build script is about 140 lines of code, yes I expect it to grow. But it will not be interacting with legacy build systems.

> 
> So like I suggested: let's first evaluate if this is really necessary. Why does druntime have to be combined with Phobos? I always found it silly that I can't just clone the dmd repo, build it, and run its test: I also have to clone Phobos, build that, just so the tests have a druntime lib to link to.
> 
> Then if we decide to keep that complexity: let's identify what the exact command is that our custom build system would run to achieve it. Is that really something dub can't with the "sourceFiles" or "postBuildCommand" directives, or do we just assume dub is too limited?
> 
> If that doesn't work, perhaps we can have a custom script for that specific complex task, without re-inventing all the rest.

Please see my comment above for a plan on how to handle this.

**Comment by thewilsonator** [comment]
> Why does druntime have to be combined with Phobos?

Also how is it supposed to be? Phobos 3 is to be source, druntime is compiled. Which should mean the only reason Phobos 3 needs to be compiled, should be to run the tests, no?

**Comment by dkorpel** [comment]
> The current build script is about 140 lines of code, yes I expect it to grow. But it will not be interacting with legacy build systems.

But currently it only implements the equivalent of a 10-line dub recipe. The whole premise of this script is that Phobos' build has to be complex for some reason, and because dub can't handle complex builds very well, the script needs to exist. It's younger than dmd's build script so doesn't have complexity related to legacy yet, but with the current mindset, it's headed to reach 1000+ lines as well.

I'm not going to block this PR, it might be necessary as a stop gap solution. I just implore you to keep Phobos v3 straightforward to build as much as possible. I believe it's not only "nice to tell someone that they can just run `dub build` or `dub run`", it's lowering the barrier to contribution when people can hack on it just as easy as any other library on dub.

**Comment by ljmf00** [comment]
> is superior to build scripts and make files.

I didn't say its superior. If the build of Phobos was a little more complex, I would probably suffer and go for Makefiles, because dub is indeed crappy for the complexity that Makefiles can achieve, but this script is a downgrade to the current situation. What if you want to build with LDC? It's flawed, only works with DMD, as is. What if you want to add flags, you have to write to your disk? It doesn't make sense.

> Like ljmf00 said, it's not about dub being particularly good, it's about complex build scripts being terrible. I don't know if you had the pleasure of dealing with [dmd's build.d](https://github.com/dlang/dmd/blob/master/compiler/src/build.d), but I certainly have.

Yes, and I also dislike the `dmd/build.d` too, idk why we would think its a great idea to reinvent this wheel many many times (which is btw a tendentious thing among devs, apparently).

**Comment by jmdavis** [comment]
> I'm not going to block this PR, it might be necessary as a stop gap solution. I just implore you to keep Phobos v3 straightforward to build as much as possible. I believe it's not only "nice to tell someone that they can just run `dub build` or `dub run`", it's lowering the barrier to contribution when people can hack on it just as easy as any other library on dub.

I have no problem whatsoever with looking at rearranging how our build system in general works. It's been the case for ages that we've had interdependencies between repos (though that was reduced by merging the dmd and druntime repos), and that's always been annoying (it also gets particularly complex when it comes to stuff like building dlang.org, since what it needs is spread across at least three repos). And I don't think that we necessarily need to keep having druntime and Phobos be in the same library file. But I'm highly skeptical that dub is a good solution for any of this given how simplistic its build system is. It's always a pain whenever you need to do anything with it that isn't absolutely bog-standard, and it's not at all designed for dealing with ancillary stuff like building the documentation. If someone wants to go to the effort of reworking all of the build stuff to use dub, then we can look at that, but I'd expect it to be pretty hideous, and a lot of it would just have to be farmed out to scripts anyway. Honestly, between make and dub, I'd rather deal with make if any build directives are necessary (though I'd rather use a D build script than either of those). Either way, we'll need to rework stuff on some level when we finally integrate the Phobos v2 and v3 builds so that v3 can be released in whatever fashion that ultimately is, and whatever we have here is clearly not that final form.

Regardless, the build system is the kind of thing that we should be able to rework as appropriate without affecting users. So, if we come up with a better solution later, we can use that then, whatever it is. For now, we need something that works, and dub doesn't.

> > Why does druntime have to be combined with Phobos?
> 
> Also how is it supposed to be? Phobos 3 is to be source, druntime is compiled. Which should mean the only reason Phobos 3 needs to be compiled, should be to run the tests, no?

The Phobos v2 build currently is combined with druntime, and at present, that continues to be the plan (though I think that there's definitely an argument for changing that). And the Phobos v2 and v3 builds will eventually need to be merged in some fashion (though how exactly that should be done is going to depend on some decisions that haven't fully been sorted out yet). At present, the only integration that we'll need is to make it so that the makefile runs the v3 unit tests so that the CI tests the v3 code.

Also, while Walter did initially say that he wanted Phobos v3 to be source-only, a number of issues have been pointed out with that, and it's highly likely that at least some of Phobos v3 will not be source-only, and it's an open question as to whether we ultimately go source-only at all. Either way, I would guess though that any new libraries we produce will not be in the same library as druntime or Phobos v2, since there should be no need for that, and we're apparently running dangerously close to the limit on the number of symbols that we can have in the Phobos library file on Windows (which would be another reason to consider splitting druntime and Phobos v2 into separate library files).

**Review by jmdavis — APPROVED** [review]

**Comment by LightBender** [comment]
> I didn't say its superior. If the build of Phobos was a little more complex, I would probably suffer and go for Makefiles, because dub is indeed crappy for the complexity that Makefiles can achieve, but this script is a downgrade to the current situation. What if you want to build with LDC? It's flawed, only works with DMD, as is. What if you want to add flags, you have to write to your disk? It doesn't make sense.

If LDC support is needed it can be added in a later pull. This PR is primarily about unblocking testing, and I'd like to keep it as as small as possible.

**Comment by Herringway** [comment]
So all these showstopper issues with dub have been filed for later fixing, right?

**Pull request merged**
